File: android/app/src/main/java/com/raceteamiq/watch/PendingSessionStore.kt
package com.raceteamiq.watch

import android.content.Context
import android.util.Log
import com.google.gson.Gson
import com.google.gson.reflect.TypeToken
import java.io.File
import java.lang.reflect.Type

private const val PENDING_FILE = "pending_sessions.json"
private val listType: Type = object : TypeToken<List<Map<String, Any?>>>() {}.type
private val gson = Gson()

/**
 * Persists session payloads so they can be sent to the phone when connected.
 * Same idea as iOS PendingSessionStore: enqueue when recording stops, remove when delivery is confirmed.
 */
class PendingSessionStore(context: Context) {
    private val file = File(context.filesDir, PENDING_FILE)

    fun enqueue(payload: Map<String, Any?>) {
        val list = allPending().toMutableList()
        list.add(payload)
        save(list)
    }

    fun remove(sessionId: String) {
        val list = allPending().filter { it["sessionId"] != sessionId }
        save(list)
    }

    fun allPending(): List<Map<String, Any?>> {
        if (!file.exists()) return emptyList()
        return try {
            val json = file.readText()
            gson.fromJson<List<Map<String, Any?>>>(json, listType) ?: emptyList()
        } catch (e: Exception) {
            Log.e("PendingSessionStore", "Failed to read pending sessions", e)
            emptyList()
        }
    }

    fun removeAll() {
        save(emptyList())
    }

    private fun save(list: List<Map<String, Any?>>) {
        try {
            file.writeText(gson.toJson(list))
        } catch (e: Exception) {
            Log.e("PendingSessionStore", "Failed to save pending sessions", e)
        }
    }
}

File: ios/RaceTeamIQWatch/Services/PendingSessionStore.swift
//
//  PendingSessionStore.swift
//  Persists session payloads so they can be sent when the watch reconnects to the phone.
//

import Foundation

/// Stores session payloads to disk so recording works offline and uploads when connected.
final class PendingSessionStore {
  static let shared = PendingSessionStore()
  
  private let fileManager = FileManager.default
  private let queueKey = "pendingSessions"
  
  private var pendingFileURL: URL? {
    guard let dir = fileManager.urls(for: .documentDirectory, in: .userDomainMask).first else { return nil }
    return dir.appendingPathComponent("pending_sessions.json", isDirectory: false)
  }
  
  private init() {}
  
  /// Append a session payload (same shape as sent to phone) for later transfer.
  func enqueue(_ payload: [String: Any]) {
    guard let url = pendingFileURL else { return }
    var list = loadPending(from: url)
    list.append(payload)
    savePending(list, to: url)
  }
  
  /// Remove the payload with the given sessionId from the queue.
  func remove(sessionId: String) {
    guard let url = pendingFileURL else { return }
    var list = loadPending(from: url)
    list.removeAll { ($0["sessionId"] as? String) == sessionId }
    savePending(list, to: url)
  }
  
  /// All pending payloads (e.g. to re-submit on activation).
  func allPending() -> [[String: Any]] {
    guard let url = pendingFileURL else { return [] }
    return loadPending(from: url)
  }
  
  /// Remove all pending (e.g. after successful transfer).
  func removeAll() {
    guard let url = pendingFileURL else { return }
    savePending([], to: url)
  }
  
  private func loadPending(from url: URL) -> [[String: Any]] {
    guard fileManager.fileExists(atPath: url.path),
          let data = try? Data(contentsOf: url),
          let json = try? JSONSerialization.jsonObject(with: data) as? [[String: Any]] else {
      return []
    }
    return json
  }
  
  private func savePending(_ list: [[String: Any]], to url: URL) {
    guard let data = try? JSONSerialization.data(withJSONObject: list) else { return }
    try? data.write(to: url)
  }
}
